### PR TITLE
fix: change deployment image reconciling

### LINKED ISSUE #346: [Bug]: Wrong deployment image processing
### Checklist

- [X] I've searched for similar issues and couldn't find anything matching
- [X] I've included steps to reproduce the behavior

### Affected Components

- [ ] K8sGPT (CLI)
- [X] K8sGPT Operator

### K8sGPT Version

v0.3.8

### Kubernetes Version

v1.28.0

### Host OS and its Version

_No response_

### Steps to reproduce

Include port in repository in `K8sGPT` object like `<host>:<port>/<path>/<image>`.

### Expected behaviour

Operator will create a deployment with normal image.

### Actual behaviour

Operator will turn image in `<host>:<tag>`, thus result into `ImagePullBackOff`.

### Additional Information

I've found the lines were this happens:
https://github.com/k8sgpt-ai/k8sgpt-operator/blob/db04fbdf23bdeeaa1053af1f578a7f3bcf7e0ca0/controllers/k8sgpt_controller.go#L168-L180

Quick solution is to parse image like this:
```go
image := strings.Split(imageURI, ":")
if len(image) == 2 {
	imageRepository = image[0]
	imageVersion = image[1]
} else {
	imageRepository = fmt.Sprintf("%s:%s", image[0], image[1])
	imageVersion = image[2]
}
```
or
```go
image := strings.Split(imageURI, ":")

imageRepository := strings.Join(image[0:len(image)-1], ":")
imageVersion := image[len(image)-1]
```

Possible better solution is to use maybe `net/url` or some other lib.

Test: https://go.dev/play/p/nIFGk2QroT5

### PR DESCRIPTION
**Pull request by ultram4rine**
Closes #346 

## 📑 Description
Fix deployment image update if repository has a port.

## ✅ Checks

- [x] My pull request adheres to the code style of this project
- [ ] My code requires changes to the documentation
- [ ] I have updated the documentation as required
- [x] All the tests have passed

## ℹ Additional Information
Also change `Update the deployment image` behavior. Now it will keep deployment image in sync with repo and version from `K8sGPT` CRD. Please tell me if this is wrong.

### CODE REVIEW

File: controllers/k8sgpt_controller.go [hunk 1]
```diff
@@ -179,18 +179,21 @@ func (r *K8sGPTReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
...
+		imageVersion := image[len(image)-1]
 
 		// if one of repository or tag is changed, we need to update the deployment
 		if imageRepository != k8sgptConfig.Spec.Repository || imageVersion != k8sgptConfig.Spec.Version {
 			// Update the deployment image
-			deployment.Spec.Template.Spec.Containers[0].Image = fmt.Sprintf("%s:%s",
-				imageRepository, k8sgptConfig.Spec.Version)
+			deployment.Spec.Template.Spec.Containers[0].Image = fmt.Sprintf(
+				"%s:%s",
+				k8sgptConfig.Spec.Repository,
```
Comment (arbreezy) [inline controllers/k8sgpt_controller.go:194]: nice catch

File: controllers/k8sgpt_controller.go [hunk 2]
```diff
@@ -179,18 +179,21 @@ func (r *K8sGPTReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
...
 	if deployment.Status.ReadyReplicas > 0 {
 
-		// Check the version of the deployment image matches the version set in the K8sGPT CR
+		// Check the repo and version of the deployment image matches the repo and version set in the K8sGPT CR
 		imageURI := deployment.Spec.Template.Spec.Containers[0].Image
 
 		image := strings.Split(imageURI, ":")
-		imageRepository := image[0]
-		imageVersion := image[1]
+		imageRepository := strings.Join(image[0:len(image)-1], ":")
```
Comment (arbreezy) [inline controllers/k8sgpt_controller.go:186]: if you don't specify a tag in your custom registry e.g `"host:8080/path/image"` the split will be wrong, right ?

**Review by arbreezy — CHANGES_REQUESTED** [review]

File: controllers/k8sgpt_controller.go (hunk 2 above)
Comment (ultram4rine) [reply controllers/k8sgpt_controller.go:186]: If I don't specify a tag code will `panic: runtime error: index out of range [1] with length 1` here I believe:
https://github.com/k8sgpt-ai/k8sgpt-operator/blob/2eae459132cec106a7b7933e419f91269e9fd8eb/controllers/k8sgpt_controller.go#L187

I will fix that somehow.

Comment (ultram4rine) [reply controllers/k8sgpt_controller.go:186]: It's about how it works now. What tag should we use it is not specified? `:latest`?

Comment (arbreezy) [reply controllers/k8sgpt_controller.go:186]: Yes, latest makes sense

**Comment by ultram4rine** [comment]
New func should work, playground: https://go.dev/play/p/eYTOsZhlJGH

**Review by arbreezy — APPROVED** [review]
more readable as well, thanks @ultram4rine

**Pull request merged**